### DISCUSSION TITLE
Building a StringRecord from a raw line

### DISCUSSION THREAD

**Discussion by Ten0** [BurntSushi/rust-csv#326] [Q&A] [answered]
It seems that currently there's no API to serde-deserialize raw records that are retrieved individually.

My use-case is basically something like:
```rust
#[derive(Deserialize)]
struct S<'a> {
    a: &'a str,
    b: &'a str,
}

let header: StringRecord = ["a", "b"].into_iter().collect(); // Constant header
let mut raw_record = csv::StringRecord::new(); // Reuse across iterations to avoid allocating

while let Some(record_str) = stream.await { // Note how this doesn't fit with Reader
    raw_record.read_raw(record_str); // This doesn't exist (question is how to set settings of course...)
    let deserialized: S = raw_record.deserialize(Some(&headers))?;
    // And do stuff with this
}
```

**Comment by Ten0** [comment]
Possible workaround:
```rust
#[tokio::main]
async fn main() {
	tokio::task::spawn(async {
		let mut csv_reader: csv::Reader<Cursor<Box<[u8]>>> = csv::ReaderBuilder::new()
			.has_headers(false)
			.from_reader(Cursor::new(Box::from([])));
		let mut string_record = csv::StringRecord::new();
		let headers = csv::StringRecord::from(vec!["a", "b", "c"]);

		let mut lines = stream::iter(["a,b,c", "d,e,f", "g,h,i"]);
		while let Some(l) = lines.next().await {
			*csv_reader.get_mut() = Cursor::new(Box::from(l.as_bytes()));
			#[derive(serde_derive::Deserialize, Debug)]
			struct S<'a> {
				a: &'a str,
				b: &'a str,
				c: &'a str,
			}
			assert!(csv_reader.read_record(&mut string_record)?);
			let deserialized: S = string_record.deserialize(Some(&headers))?;
			dbg!(deserialized);
		}
		Ok::<_, anyhow::Error>(())
	}).unwrap();
}
```

This seems pretty dirty: it seems pretty uncertain that this won't break on future `csv` updates.

Also, it uses the internal reader's BufReader as intermediate buffer, which creates unnecessary additional copies -> overhead.

- Is it guaranteed that this would work across `csv` releases? (If so, that should probably be documented on `get_mut()`, and we can add the test below to the test suite)
- It seems that we'd be better off if we could construct the StringRecord without the extra alloc. That would be possible reasonably easily if either:
  + `ByteRecord::as_parts` was exposed (we could call `csv_core` ourselves, and construct the `ByteRecord` manually, *then* call the `Deserializer`, but that wouldn't be very ergonomic)
  + there was a `StringRecord::read_raw(&mut self, &[u8], &mut csv_core::Reader)` function (but that would imply that upgrading major versions of `csv_core` would become a breaking change for `csv`)

Test to add to the test suite if this workaround is actually guaranteed behavior:
```rust
#[test]
fn test_csv_line_by_line_feeding() {
	let mut csv_reader: csv::Reader<Cursor<Box<[u8]>>> = csv::ReaderBuilder::new()
		.has_headers(false)
		.from_reader(Cursor::new(Box::from([])));
	let mut string_record = csv::StringRecord::new();
	let headers = csv::StringRecord::from(vec!["a", "b", "c"]);

	#[derive(serde_derive::Deserialize, Debug, PartialEq, Eq)]
	struct S<'a> {
		a: &'a str,
		b: &'a str,
		c: &'a str,
	}

	let input = ["a,b,c", "d,e,f\n\naaaaa\n", "g,h,i\n"];
	let expected_out = [
		S { a: "a", b: "b", c: "c" },
		S { a: "d", b: "e", c: "f" },
		S { a: "g", b: "h", c: "i" },
	];
	assert_eq!(input.len(), expected_out.len());

	for (line, expected_out) in input.iter().zip(expected_out) {
		*csv_reader.get_mut() = Cursor::new(Box::from(line.as_bytes()));
		csv_reader.seek(csv::Position::new()).unwrap(); // This will reset the reader state
		assert!(csv_reader.read_record(&mut string_record).unwrap());
		let deserialized: S = string_record.deserialize(Some(&headers)).unwrap();
		assert_eq!(deserialized, expected_out);
	}
}
```

**Comment by BurntSushi** [comment, marked as answer]
I can't make heads or tails of what you're trying to do. But [`Reader::read_record`](https://docs.rs/csv/latest/csv/struct.Reader.html#method.read_record) is the way to read a single record.

If you want to parse a single CSV record from a single line, then you need to create a `Reader` for that single line.

But more generally, "parse a single CSV record from a single line" is incorrect because a CSV record can span multiple lines.

**Reply by Ten0** [reply]
Hey! Thanks a lot for your quick answer!

> I can't make heads or tails of what you're trying to do

Sorry for being unclear :/ I'll try to be more clear:
- I receive CSV `&[u8]`s that represent records one by one. A loooot of them. (Like 1 million/s)
- I don't want to pay the cost of initializing a new reader for each of them, because otherwise I won't hold the speed requirements. The workaround above seems to achieve that.

> more generally, "parse a single CSV record from a single line" is incorrect because a CSV record can span multiple lines.

I didn't know that, thanks ^^. (I thought newlines would be escaped as "\n")
That doesn't change much though: In that case I may receive inputs as separate messages, each of which representing a record. (no matter whether it's multi-line or not)

**Reply by BurntSushi** [reply]
Newlines are not escaped. They are just within a quoted field.

The `csv` crate doesn't support your use case well. Sorry. You might have luck dropping down to the `csv-core` crate.

**Reply by Ten0** [reply]
The issue with going through the `csv-core` crate is we can't construct the `StringRecord` cheaply from that currently, so we lose the `serde` capabilities:
- There is no dedicated `StringRecord::read_raw(&mut self, &[u8], &mut csv_core::Reader)` function
- `ByteRecord::as_parts` is private so we can't copy-paste the implementation within `csv::Reader`

Hence my previous message:
> It seems that we'd be better off if we could construct the StringRecord without the extra alloc. That would be possible reasonably easily if either:
> - ByteRecord::as_parts was exposed (we could call csv_core ourselves, and construct the ByteRecord manually, then call the Deserializer, but that wouldn't be very ergonomic)
> - there was a StringRecord::read_raw(&mut self, &[u8], &mut csv_core::Reader) function (but that would imply that upgrading major versions of csv_core would become a breaking change for csv)

**Reply by BurntSushi** [reply]
If you're using Serde then you're already paying an overhead tax. Have you benchmarked it?

Also:

> StringRecord::read_raw(&mut self, &[u8], &mut csv_core::Reader)

That is seemingly the exact same signature as `Reader::read_record` but with the order of parameters changed.

I repeat that your use case doesn't make a lot of sense to me, is not well supported by this library and I don't see any compelling reason to support it. The architecture you've described seems supremely weird, and the concerns about perf don't feel like they are well-founded to me.

**Reply by BurntSushi** [reply]
If there is a simple change that will let you do what you want with csv-core then I'm open to that. Please open a PR. I won't necessarily rubber stamp exposing private APIs if it inhibits API evolution, but we can look at a proposal in detail.

**Reply by Ten0** [reply]
> That is seemingly the exact same signature as Reader::read_record but with the order of parameters changed.

The other main difference is that it reads from a provided slice, which would allow feeding those one by one.

> If you're using Serde then you're already paying an overhead tax. Have you benchmarked it?

Yes I have, it doesn't prevent me from reaching the million lines per second, and is also a good compromise with regards to behavior evolution/code readability/performance (headers can be dynamically parametrized, but struct into which we're parsing can't). While I don't think the extra copy I pay here is going to prevent that either, it seems that it comes with no code readability benefits however, so that doesn't *feel* like an optimal solution (but it may still be the best compromise for the reasons you mentioned).

> If there is a simple change that will let you do what you want with csv-core then I'm open to that. Please open a PR. I won't necessarily rubber stamp exposing private APIs if it inhibits API evolution, but we can look at a proposal in detail.

Either of these two changes would allow avoiding this extra copy AFAICT:
- if ByteRecord::as_parts was exposed (we could call csv_core ourselves, and construct the ByteRecord manually, then call the Deserializer, but that wouldn't be very ergonomic)
- if there was a StringRecord::read_raw(&mut self, &[u8], &mut csv_core::Reader) function (but that would imply that upgrading major versions of csv_core would become a breaking change for csv)

I have no particular preference on which would be best.

However given the two rather large downsides with regards to API evolution (as you said, it seems this use-case doesn't fit very well with the `csv` crate - which is otherwise really great btw, thanks again), guaranteeing via adding [the test above](https://github.com/BurntSushi/rust-csv/discussions/326#discussioncomment-6206525) that the workaround keeps working across minor releases of `csv` feels to me like the most reasonable compromise (it's unlikely that that's going to change anyway, and at least it would guarantee that it would be done consciously). (That's still one more copy than would be necessary, through the BufRead, but that doesn't seem like a *big* issue.)

Should I open a PR for this?

**Reply by BurntSushi** [reply]
> The other main difference is that it reads from a provided slice, which would allow feeding those one by one.

I still think you should do this:

```rust
fn main() {
    let raw = r"foo,bar,baz";
    let mut sr = csv::StringRecord::new();
    csv::ReaderBuilder::new()
        .has_headers(false)
        .from_reader(raw.as_bytes())
        .read_record(&mut sr)
        .unwrap();
    dbg!(&sr);
}
```

You've presumably concluded that you don't want the overhead of building a reader, but it's not clear to me it's actually significant, especially if you're using Serde. You say it is though. I guess I have to take your word for it. Maybe there is a way to make building a `csv::Reader` faster.

> if `ByteRecord::as_parts` was exposed

Okay, so looking at this, I don't think that's doable. It exposes the internal implementation details of a `ByteRecord` and pins us to using that for the future. (Because it means we have to document what the two parts actually mean.) And if something like an `as_parts` is exposed, then we probably need a `from_parts` too. It's not a simple change and does indeed inhibit API evolution.

> if there was a StringRecord::read_raw(&mut self, &[u8], &mut csv_core::Reader) function (but that would imply that upgrading major versions of csv_core would become a breaking change for csv)

This is also definitely not going to happen. There's no way I'm ever going to make `csv-core` a public dependency of `csv`. It is very specifically not a public dependency, by design.

> guaranteeing via adding https://github.com/BurntSushi/rust-csv/discussions/326#discussioncomment-6206525 that the workaround keeps working across minor releases of csv feels to me like the most reasonable compromise (it's unlikely that that's going to change anyway, and at least it would guarantee that it would be done consciously).

Changing the underlying reader out from underneath the `csv::Reader` is maybe going to work some of the time, but I don't think I can comfortably commit to guaranteeing that. There's a whole bunch of stateful shit happening inside a `csv::Reader`. It is not designed to have its underlying reader swapped out like that. It's *probably* the case that it will keep working in very limited cases, but it doesn't really look like an API guarantee that I want to target.

The other alternative here is to just use `csv-core` and roll your own Serde stuff. It's undoubtedly a huge pain unfortunately. The Serde stuff inside the `csv` crate is a monstrous nightmare and has a number of problems in and of itself. But if you can limit the Serde support to what you need specifically, you might be able to significantly simplify it.

The whole point of `csv-core` is that, while in some case inconvenient, you *can* roll all your own stuff if you need to without actually re-implementing a CSV parser. That's the reason I split them apart because I knew the higher level design of `csv` wouldn't and couldn't be all things to all people. I think for cases like this, you're just going to have to come to terms with either some kind of performance overhead or some development overhead.

**Reply by BurntSushi** [reply]
One other alternative in the design space here is to split apart "CSV reader configuration" and "CSV reader building." Then you could expose a `StringReader::read_raw(&mut self, config: &mut csv::Config, data: &[u8])`. But that's a fairly sizeable API addition and I'm not sure it's worth doing just for cases like this.

**Reply by Ten0** [reply]
> One other alternative in the design space here is to split apart "CSV reader configuration" and "CSV reader building."

Yeah I've thought about that option as well, and figured that it seems that would mean refactoring a big part of the reader out of itself into Config. I'm not sure that would be worth doing either for such a niche use-case. (If you happen to have time to review such big changes it would probably be better spent unblocking the serde flatten support PR 😅)

> It's probably the case that it will keep working in very limited cases, but it doesn't really look like an API guarantee that I want to target.

It seems that given the current signatures of csv_core it could hardly behave differently.
How about just adding the test then, without guaranteeing it, with a comment that says if it's required we can remove compatibility, but at least that wouldn't happen accidentally if it's easy to maintain support?

The advantage of going through the Reader is that checks such as "flexible" are maintained, and code is simpler, compared to the version where one would manually populate the StringRecord using csv_core.
